File: python/handlers/transactions.py
"""
Transaction handlers (IPC-only).

Group a sequence of mutating MCP calls into a single KiCad undo step.
Without this, every place_component / route_trace / add_zone call ends
up as its own entry in the undo history — an AI agent that does a
five-step refactor leaves five Ctrl-Z steps for the human to walk
back through.  With ``begin_transaction`` open, those five calls
collapse into one.

Caveat: only mutations that go through the board's create_items /
update_items / remove_items path participate.  Property mutations like
``set_origin`` and ``set_title_block_info`` are sent as direct kipy
commands and do NOT join the transaction (kipy treats them as
out-of-band — they apply immediately and are individually undoable).
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any, Dict

if TYPE_CHECKING:
    from kicad_interface import KiCADInterface

logger = logging.getLogger(__name__)


def _ipc_unavailable(reason: str = "") -> Dict[str, Any]:
    base = (
        "Transaction commands require the IPC backend. Launch KiCAD "
        "with Preferences > Plugins > Enable IPC API Server, then retry."
    )
    return {"success": False, "message": f"{base} ({reason})" if reason else base}


def _require_ipc(iface: "KiCADInterface") -> Dict[str, Any]:
    if iface.use_ipc and iface.ipc_board_api:
        return {}
    ok, reason = iface.ensure_ipc(allow_launch=True)
    if ok:
        return {}
    return _ipc_unavailable(reason)


def handle_begin_transaction(
    iface: "KiCADInterface", params: Dict[str, Any]
) -> Dict[str, Any]:
    """Open a transaction. Refuses to nest — commit or rollback first.

    Forwards ``description`` as-is (including explicit empty string).
    The backend supplies the default label when the key is omitted /
    null; this keeps the "MCP Operation" default in one place.
    """
    gate = _require_ipc(iface)
    if gate:
        return gate
    description = params.get("description")
    if description is not None:
        description = str(description)
    return iface.ipc_board_api.begin_transaction(description)


def handle_commit_transaction(
    iface: "KiCADInterface", params: Dict[str, Any]
) -> Dict[str, Any]:
    """Push the open transaction as one undo step.

    ``description`` overrides the label set at begin_transaction; omit it
    to keep the original label.
    """
    gate = _require_ipc(iface)
    if gate:
        return gate
    description = params.get("description")
    if description is not None:
        description = str(description)
    return iface.ipc_board_api.commit_transaction(description)


def handle_rollback_transaction(
    iface: "KiCADInterface", params: Dict[str, Any]
) -> Dict[str, Any]:
    """Discard the open transaction — every change since begin is reverted."""
    gate = _require_ipc(iface)
    if gate:
        return gate
    return iface.ipc_board_api.rollback_transaction()


def handle_get_transaction_status(
    iface: "KiCADInterface", params: Dict[str, Any]
) -> Dict[str, Any]:
    """Report whether a transaction is currently open."""
    gate = _require_ipc(iface)
    if gate:
        return gate
    return iface.ipc_board_api.get_transaction_status()

File: src/tools/transactions.ts
/**
 * Transaction tools (IPC-only).
 *
 * Group a sequence of mutating MCP calls into a single KiCad undo step.
 * Useful for multi-step agentic workflows — an AI doing a 5-step PCB
 * refactor produces ONE Ctrl-Z entry the human can use to bail out,
 * not five.
 *
 * Workflow:
 *   1. transaction({action: "begin", description: "Move power section"})
 *   2. move_component / route_trace / add_via / ...   (any number)
 *   3. transaction({action: "commit"}) — atomic undo step lands
 *
 * If anything fails partway through, call transaction({action: "rollback"})
 * to discard everything since begin.
 *
 * Caveat: only create / update / remove of board items participate.
 * set_origin and set_title_block_info are sent as direct kipy property
 * commands and apply immediately, outside the transaction.
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { logger } from "../logger.js";
import { CommandFunction, formatKicadResult } from "./tool-response.js";

export function registerTransactionTools(server: McpServer, callKicadScript: CommandFunction) {
  server.tool(
    "transaction",
    "Manage a KiCad transaction / undo group (IPC-only). 'begin' opens one so subsequent mutating calls collapse into a single undo step (no nesting); 'commit' lands it atomically; 'rollback' discards all changes since begin; 'status' reports the open transaction.",
    {
      action: z
        .enum(["begin", "commit", "rollback", "status"])
        .describe("begin | commit | rollback | status"),
      description: z
        .string()
        .optional()
        .describe("Undo label: 'begin' sets it (default 'MCP Operation'); 'commit' overrides it"),
    },
    async (args: { action: "begin" | "commit" | "rollback" | "status"; description?: string }) => {
      const commandByAction = {
        begin: "begin_transaction",
        commit: "commit_transaction",
        rollback: "rollback_transaction",
        status: "get_transaction_status",
      } as const;
      const { action, ...rest } = args;
      const result = await callKicadScript(commandByAction[action], rest);
      return formatKicadResult(result);
    },
  );

  logger.info("Transaction tools registered (1 tool)");
}
